packager/mux/mux_test.go

```diff
@@ -0,0 +1,199 @@
+	templateURL = strings.Replace(templateURL, "$CERT", pkgt.CertName, 1)
+	return templateURL
+}
+
+// Dedicated type for annotated params - for tests readability.
+type params map[string]string
+
+// Mock for underlying http handlers - signer, cert etc.
+type MockedHandler struct {
+	mock.Mock
+}
+
+// Mock ServeHTTP: record annotated params, don't forward the call.
+func (m *MockedHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	// Convert annotated parameters to dedicated "params" type for tests readability.
+	m.Called(params(Params(req)))
+}
+
+func TestServeHTTPSuccess(t *testing.T) {
+	tests := []struct {
+		expectedHandlerName  string
+		testScenario         string
+		url                  string
+		expectedParsedParams params
+	}{
+		{"signer",
+			" empty",
+			Expand("$HOST/priv/doc"),
+			params{}},
+		{"signer",
+			" with query, empty",
+			Expand("$HOST/priv/doc?"),
+			params{}},
+		{"signer",
+			" with query, regular",
+			Expand("$HOST/priv/doc?fetch=$FETCH&sign=$SIGN"),
+			params{}},
+		{"signer",
+			" with query, escaping",
+			Expand("$HOST/priv/doc?fetch=$FETCH&sign=$SIGN%2A\\"),
+			params{}},
+		{"signer",
+			" with path, empty",
+			Expand("$HOST/priv/doc/"),
+			params{"signURL": ""}},
+		{"signer",
+			"  with path, regular",
+			Expand("$HOST/priv/doc/$FETCH"),
+			params{"signURL": Expand("$FETCH")}},
+		{"signer",
+			" with path, escaping",
+			Expand("$HOST/priv/doc/$FETCH%2A\\"),
+			params{"signURL": Expand("$FETCH%2A%5C")}},
+		{"signer",
+			" with path and query, regular",
+			Expand("$HOST/priv/doc/$FETCH?amp=1"),
+			params{"signURL": Expand("$FETCH?amp=1")}},
+		{"signer",
+			" with path and query, escaping",
+			Expand("$HOST/priv/doc/$FETCH%2A\\?amp=1%2A\\"),
+			params{"signURL": Expand("$FETCH%2A%5C?amp=1%2A\\")}},
+		{"cert",
+			" empty",
+			Expand("$HOST/amppkg/cert/"),
+			params{"certName": ""}},
+		{"cert",
+			" regular",
+			Expand("$HOST/amppkg/cert/$CERT"),
+			params{"certName": Expand("$CERT")}},
+		{"cert",
+			" escaping",
+			Expand("$HOST/amppkg/cert/$CERT%2A\\"),
+			params{"certName": Expand("$CERT*\\")}},
+		{"validityMap",
+			" regular",
+			Expand("$HOST/amppkg/validity"),
+			params{}},
+		{"healthz",
+			" regular",
+			Expand("$HOST/healthz"),
+			params{}},
+	}
+	for _, tt := range tests {
+		testName := tt.expectedHandlerName + tt.testScenario
+		t.Run(testName, func(t *testing.T) {
+			mocks := map[string](*MockedHandler){"signer": &MockedHandler{}, "healthz": &MockedHandler{}, "cert": &MockedHandler{}, "validityMap": &MockedHandler{}}
+			mux := New(mocks["cert"], mocks["signer"], mocks["validityMap"], mocks["healthz"])
+
+			// Set expectation.
+			expectedMockedHandler := mocks[tt.expectedHandlerName]
+			expectedMockedHandler.On("ServeHTTP", tt.expectedParsedParams)
+
+			// Run.
+			actualResp := pkgt.Get(t, mux, tt.url)
+
+			// Expect no errors.
+			assert.Equal(t, 200, actualResp.StatusCode, "No error expected: %#v", actualResp)
```

Comment:
optional: I learned this when my changes for another project got reviewed by another team that we shouldn't be using assert in unit tests for go, so it's ok if you want to keep it this way.
go/gocomments#assert

Reply:
Hm, I've looked into this, and also the relevant [FAQ](https://golang.org/doc/faq#assertions), and looks like asserts are actually more helpful than the go's native way.

My thoughts on go's FAQ:

> Go doesn't provide assertions. They are undeniably convenient, but our experience has been that programmers use them as a crutch to avoid thinking about proper error handling and reporting. Proper error handling means that servers continue to operate instead of crashing after a non-fatal error.

For tests it's actually helpful to terminate the test automatically on first failed assertion. E.g. if a functional test parses a URL from XML and then fetches it, I'd like the test to stop if parsing failed - instead of trying to fetch a broken URL. So assertion is kind of like throwing an unhandled exception, with is acceptable and convenient for tests IMO.

> Proper error reporting means that errors are direct and to the point, saving the programmer from interpreting a large crash trace. Precise errors are particularly important when the programmer seeing the errors is not familiar with the code.

I agree. Yet golang's native way requires more work to implement error reporting than testify/assert package. 

testify/assert provides precise reporting for free. When an assertion fails, the framework provides the test name, the type of musmatch (e.g. "not equal"), the mismatched values, the diff, and the line number:

```
--- FAIL: TestParamsIncorrectValueType (0.00s)
	mux_test.go:209: 
			Error Trace:	mux_test.go:209
			Error:      	Not equal: 
			            	expected: map[string]string{}
			            	actual  : map[string]string{"":""}
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1,2 +1,3 @@
			            	-(map[string]string) {
			            	+(map[string]string) (len=1) {
			            	+ (string) "": (string) ""
			            	 }
			Test:       	TestParamsIncorrectValueType
FAIL
```

On the contrary, the native "t.Error" output looks like this:

```
--- FAIL: TestParamsIncorrectValueType (0.00s)
	mux_test.go:210: 
FAIL
```

Reply:
I am also on the side of using asserts, that is why I marked this comment as optional. I am actually not sure why the go style guide promotes non-use of asserts but I'm sure it's not without controversy.

Reply:
I'm also on the side of keeping the asserts! Had this comment sitting in draft:

I was unconvinced by the reasoning behind that go/ link (and since this codebase lives outside of Google, I don't think we're beholden to that doc). My reasoning for using the assert lib everywhere is that it improves readability, by removing boilerplate and letting the reader focus on business logic.